File: 백준/Gold/9663. N－Queen/N－Queen.py
N = int(input())
hori = [False]*N
diag1 = [False]*(2*N-1)
diag2 = [False]*(2*N-1)

def check(i,j):
    hori[j], diag1[i+j], diag2[N-1-i+j] = True,True,True
def un_check(i,j):
    hori[j], diag1[i+j], diag2[N-1-i+j] = False,False,False

ans = 0
def dfs(i):
    global ans
    if i == N-1: 
        ans += 1
        return
    
    i2=i+1
    for j2 in range(N):
        if hori[j2] or diag1[i2+j2] or diag2[N-1-i2+j2]: continue
        
        check(i2,j2)
        dfs(i2)
        un_check(i2,j2)
dfs(-1)
print(ans)

File: 프로그래머스/2/12952. N－Queen/N－Queen.js
let rows, cols;
let N;
let cnt = 0;
let diag1, diag2;
function dfs(i){
    if (i=== N) {
        cnt += 1;
        return;
    }
    
    for (let j=0; j<N; j++){
        if (cols[j] || diag1[i+j] || diag2[i+(N-1-j)]) continue;
        cols[j] = true; diag1[i+j] = true; diag2[i+(N-1-j)] = true;
        dfs(i+1);
        [cols[j], diag1[i+j], diag2[i+(N-1-j)]] = [false, false,false];
    }
}

function solution(n) {
    N=n;
    cols = Array.from({length:N}, ()=>false);
    diag1 = Array.from({length:2*N-1}, ()=>false);
    diag2 = Array.from({length:2*N-1}, ()=>false);
    dfs(0)

    return cnt;
}